Result/Result.swift

```diff
@@ -31,6 +31,15 @@ public enum Result<T, Error: ErrorType>: ResultType, CustomStringConvertible, Cu
 		}
 	}
 	
+	/// Constructs a result from a function that uses `throw`, failing with `Error` if throws, non-autoclosure version
+    	public init?(_ f: () throws -> T?) {
+        	do {
+            		guard let result = try f() else { return nil }
```

Comment:
I'm confused I think this should be defined in terms of `T`, not `T?`?

```swift
    	/// Constructs a result from a function that uses `throw`, failing with `Error` if throws, non-autoclosure version
    	public init(_ f: () throws -> T) {
        	do {
            		self = .Success(try f())
		} catch {
        		self = .Failure(error as! Error)
        	}
        }
```

Then I think the `@autoclosure` version can be reimplemented based on this one :)